### ISSUE TITLE
Kathy Rain: Unable to zoom into Picture (Day 2)

### ISSUE THREAD

**Issue by Gemba** [adventuregamestudio/ags#1936]
**Describe the bug**
On Day 2 in the dorm it is not possible to zoom into the reconstructed picture with the _Image Analyzer_ tool at Eleen's computer.

**AGS Version**
Source build from v.3.6.0.43, 8dce503b

**Game**
Kathy Rain, GOG, v1.0.4

**To Reproduce**
1. Pre-Req: Have the overexposed picture on Eleen's computer reconstructed, printed out and analyzed.
    1. Use Computer, know the credentials
    2. Click on "submit" button on login screen;
    3. Drag _bright_picture.bmp_ file icon onto the _Image Analyzer_ program icon.
    4. Move sliders until you get a bright colored image.
    5. Print it and look/think on the picture in your inventory.
2. Use the _Image Analyzer_ tool on her computer again to zoom into the picture (with the <+> cursor)
3. The picture disappears and only the buttons of the image tool (_Return_, _Analyze_, _Printout_) are available.
4. With _Return_ the Image tool can be closed and then restarted.
5. No crash, but this issue hinders game completion.

To reconstruct, see also attached saved game. [agssave.004.gz](https://github.com/adventuregamestudio/ags/files/10837026/agssave.004.gz)

**Expected behavior**
Clicking on the picture with the <+> cursor should zoom into some region of the picture and the picture should stay in the foreground right below the controls of the image tool, i.e. the picture should not "disappear" to a lower z-layer. See also screens attached.

**Screenshots**

Before zoom:

![01_before_zoom_click](https://user-images.githubusercontent.com/418292/221498985-adf2c4c6-d941-4cb3-8399-47754b361c4f.png)

After trying to zoom:

![02_after_zoom_click](https://user-images.githubusercontent.com/418292/221498989-dfb6d04b-bea4-45f9-918b-9eaa04862a31.png)

**Desktop:**
 - OS: Linux, Raspi OS, Buster, packages updated yesterday.
 - Version: `Linux rpi4 4.19.97-v7l+ #1294 SMP Thu Jan 30 13:21:14 GMT 2020 armv7l GNU/Linux`, `libsdl2-dev_2.0.10`

**Additional context**
Directly run on fKMS graphic context via SDL2, no X11 used.

**Comment by ivan-mogilko** [comment]
Good thing is that GOG save also works with a Steam version (which i happen to have).

More elaborate first steps:
- use Computer;
- click on "submit" button on login screen;
- drag "bright_picture.bmp" file icon onto the "Image Analyzer" program icon.

**Comment by Gemba** [comment]
Thanks for the elaborated actions on how to reproduce. I added them to the initial issue description.

I also added a saved game from the Day 2 start ([agssave.002.gz](https://github.com/adventuregamestudio/ags/files/10875424/agssave.002.gz)), if it helps.

I get this information reported on the `Kathy Rain.exe` when running `ags`, apparently it was compiled with ags 3.3.0.x:
```
[...]
Opened game data file: game28.dta
Game data version: 43
Compiled with: 3.3.0.1150
[...]
```

**Comment by Gemba** [comment]
Update to this: It seems the zoomed in parts are not displayed but logicwise the game works.

Workaround:
When looking at the picture remember the positions which you want to zoom (there are two objects on the picture which stick out). Click on the area once, and then a second time until the <-> cursor appears. Then press the button _Analyze Image_. If you hit the right position Kathy will comment on that. After the two objects are analyzed the game can be continued.

From the distance I assume it is either related to the zooming/scaling and/or z-layer ordering.

**Comment by ivan-mogilko** [comment]
> 
> I also added a saved game from the Day 2 start ([agssave.002.gz](https://github.com/adventuregamestudio/ags/files/10875424/agssave.002.gz)), if it helps.
> 

Sorry, what I meant, i wanted to walkthrough the original binary, to be able to compare the difference.

Also, having a save in older format makes it easier to test this with different engine builds, because earlier versions of the engine may not read latest save format.
I'll attach one made in original game here for the reference, of approx. same spot:
[agssave.zip](https://github.com/adventuregamestudio/ags/files/10886422/agssave.zip)

Unfortunately, my plan failed with this save, because there's either a different with save format ("Kathy Rain" is known to be made with a slightly changed engine, not the official ags engine), or there's some another regression in 3.6.0.
(At least this save may be used with the original binary)

I will investigate this a bit, but regardless will get back to the picture problem soon.

**Comment by ericoporto** [comment]
Actually, I think their engine for Kathy Rain is https://github.com/josthas/KrusAGS

**Comment by ivan-mogilko** [comment]
> Actually, I think their engine for Kathy Rain is https://github.com/josthas/KrusAGS

Ah, right, I forgot about that one.

**Comment by ivan-mogilko** [comment]
Sigh,

So what happens is this.
* The zooming photograph is a GUI with a DynamicSprite assigned to it as a BackgroundGraphic when the analyzer is turned on.
* When player zooms in and out the new DynamicSprite is created, acting as scaled image.
* But the newly created DynamicSprite is never reassigned to the GUI.

So why it works in the original game? Because the newly created DynamicSprite, by pure coincidence, is assigned same index as the previous one, so when the engine wants to draw gui again it does not "notice" that anything had happened.

Why it does not work with the new engine? Because the reaction to deleted dynamic sprites were expanded recently. Previously it only cleared deleted sprite's references in gui buttons and objects (reassigning to a dummy sprite). Now it removes them everywhere.

In a nutshell, this is a combination of a scripting mistake and engine behavior getting more consistent.

**Comment by ivan-mogilko** [comment]
Pushed a fix: e8409920049b2a8859856d3525a8f112227df4df

**Comment by Gemba** [comment]
Thanks for the helpful information, helped me to understand a little more how AGS works.
Also thank you for swiftly hunting this down.

I can confirm with my setup that your change fixes this issue.
